### ISSUE TITLE
How to shuffle per-item examples

### ISSUE THREAD

**Issue by johnvorsten** [tensorflow/ranking#131]
My first model was not generalizing well.  I think I'm over-fitting the data, and I think it is partially caused by my per-item examples being pre-sorted before they're fed into the model.  For example, my per-item features are being fed into the model from most relevant to least relevant and the model just learns to pick the first items.

Is there a built-in way to shuffle per-item examples?  I'm trying to understand if what I want is related to the group_size parameter, and the methods tfr.model._update_scatter_gather_indices and tfr.model._compute_logits_impl.  Do you have any reading where I can understand what group_size is/does for LTR?

If there is not built-in utility my plan is to shuffle in my score_function depending if its in training or prediction mode.

Thanks!

**Comment by johnvorsten** [comment]
I added a shuffling feature under _compute_logits_impl in _RankingModel.  Would it be a useful PR?

**Comment by HongleiZhuang** [comment]
Hi @johnvorsten ,

Can you clarify whether you are trying to shuffle features within each example, or to shuffle examples within each list (query)?  Shuffling features should not affect the performance at all, and shuffling examples within each list would not affect the performance either when `group_size=1`.

**Comment by johnvorsten** [comment]
@HongleiZhuang I'm trying to shuffle examples within each list with group_size = list_size.  I have 200 examples and I would like to get a complete list-wise ranking of those examples for a query

**Comment by HongleiZhuang** [comment]
@johnvorsten In this case, the examples should have been shuffled within each list when groups are formed.  See https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/python/model.py#L311-L334

Do you see any performance differences when you do your shuffling in `_compute_logits_impl`?

However, if your input data has more than 200 examples in each list and you set list_size to 200, the list would be truncated to the top-200 examples (without shuffling).

**Comment by johnvorsten** [comment]
@HongleiZhuang I added a name space to shuffle features where you linked. 

I tried (2) things to get my model to generalize better : 
1) Shuffle examples before I serialized data (and do not shuffle examples in my model)
2) Shuffle examples in _compute_logits_impl as suggested

I did not notice a difference in performance between these two methods.  I did see a big performance drop compared to my first model where I did not shuffle examples at all.

I think this issue can be closed.  I'll add a PR for how I added shuffling, and if y'all can decide if you think it will be a useful addition
